pulsar/producer_partition.go

```diff
@@ -542,6 +538,11 @@ func (p *partitionProducer) internalSend(request *sendRequest) {
 
 	uncompressedSize := len(uncompressedPayload)
 
+	// try to reserve memory for uncompressedPayload
+	if !p.canReserveMem(request, int64(uncompressedSize)) {
+		return
```

Comment:
forget to release semaphore and runCallback

Reply:
Semaphore is released by `canReserveMem` if it failed, `runCallback` as well.

Reply:
And I think Semaphore released by `canReserveMem ` may not be a good idea. It makes `canReserveMem` must be invoked after `canAddToQueue `. What do you think?

Reply:
I think so, runCallback/releaseSemaphore/releaseMemory in canAddToQueue and canReserveMem violates the Single Responsibility Principle

Reply:
The final solution is encapsulating the resource releasing logic in request.done(), anywhere there is an error, just call request.done()

Reply:
I'm +1 for moving the semaphore release out of `canReserveMem`. It's better that we release it here than in the `canReserveMem` before we find a good solution for it.